stats.go

```diff
@@ -450,74 +484,169 @@ func (s *statStore) NewPerInstanceTimer(name string, tags map[string]string) Tim
+		s.mergePerInstanceTags(tags))
 }
 
-func (s subScope) NewTimer(name string) Timer {
+func (s *subScope) NewTimer(name string) Timer {
 	return s.NewTimerWithTags(name, nil)
 }
 
-func (s subScope) NewTimerWithTags(name string, tags map[string]string) Timer {
-	return s.registry.NewTimerWithTags(joinScopes(s.name, name), s.mergeTags(tags))
+func (s *subScope) NewTimerWithTags(name string, tags map[string]string) Timer {
+	return s.registry.newTimerWithTagSet(joinScopes(s.name, name), s.mergeTags(tags))
 }
 
-func (s subScope) NewPerInstanceTimer(name string, tags map[string]string) Timer {
-	return s.registry.NewPerInstanceTimer(joinScopes(s.name, name), s.mergeTags(tags))
+func (s *subScope) NewPerInstanceTimer(name string, tags map[string]string) Timer {
+	return s.registry.newTimerWithTagSet(joinScopes(s.name, name),
+		s.mergePerInstanceTags(tags))
 }
 
 func joinScopes(parent, child string) string {
 	return parent + "." + child
 }
 
-// mergeTags augments tags with all scope-level tags that are not already present.
-// Modifies and returns tags directly.
-func (s subScope) mergeTags(tags map[string]string) map[string]string {
-	if len(s.tags) == 0 {
-		return tags
+// mergeOneTag is an optimized for inserting 1 tag and will panic otherwise.
+func (s *subScope) mergeOneTag(tags map[string]string) tagSet {
+	if len(tags) != 1 {
+		panic("invalid usage")
 	}
-	if len(tags) == 0 {
+	var p tagPair
+	for k, v := range tags {
+		p = tagPair{key: k, value: replaceChars(v)}
+		break
+	}
+	if p.key == "" || p.value == "" {
 		return s.tags
 	}
-	for k, v := range s.tags {
-		if _, ok := tags[k]; !ok {
-			tags[k] = v
+	a := make(tagSet, len(s.tags), len(s.tags)+1)
+	copy(a, s.tags)
+	return a.Insert(p)
+}
+
+// mergeTags returns a tagSet that is the union of subScope's tags and the
+// provided tags map. If any keys overlap the values from the provided map
+// are used.
+func (s *subScope) mergeTags(tags map[string]string) tagSet {
+	switch len(tags) {
+	case 0:
+		return s.tags
+	case 1:
+		// optimize for the common case of there only being one tag
+		return s.mergeOneTag(tags)
+	default:
+		// write tags to the end of the scratch slice
+		scratch := make(tagSet, len(s.tags)+len(tags))
+		a := scratch[len(s.tags):]
+		i := 0
+		for k, v := range tags {
+			if k != "" && v != "" {
+				a[i] = tagPair{key: k, value: replaceChars(v)}
+				i++
+			}
+		}
+		a = a[:i]
+		a.Sort()
+
+		if len(s.tags) == 0 {
+			return a
 		}
+		return mergeTagSets(s.tags, a, scratch)
+	}
+}
+
+// mergePerInstanceTags returns a tagSet that is the union of subScope's
+// tags and the provided tags map with. If any keys overlap the values from
+// the provided map are used.
+//
+// The returned tagSet will have a per-instance key ("_f") and if neither the
+// subScope or tags have this key it's value will be the default per-instance
+// value ("i").
+//
+// The method does not optimize for the case where there is only one tag
```

Comment:
Seems like it would be possible to check the length of `tags` and have 2 impls, one for one tag and one for many?

Reply:
Honestly, the one tag optimization is only worth is when `s.pairs` is large (4+ elements) and we're only talking about a 100-400ns in savings. The optimization here would require a lot of duplicated logic to check for the `"_f"` tag and is not worth the added complexity in more opinion (and I lean towards speed over readability).